File: huannv_Rx_Retrofit_MultiThread/app/src/main/java/com/example/base_project/base/BaseActivity.java
package com.example.base_project.base;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.ViewModelProvider;

import com.example.base_project.R;

import butterknife.ButterKnife;

public abstract class BaseActivity extends AppCompatActivity {
    private FragmentTransaction transaction;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getLayoutId());
        ButterKnife.bind(this);
        transaction = getSupportFragmentManager().beginTransaction();
        initAct();
    }

    protected int getLayoutId() {
        return R.layout.layout_container;
    }

    protected abstract void initAct();

    protected void pushView(Fragment fragment) {
        transaction.add(R.id.container_frame, fragment);
        transaction.commitAllowingStateLoss();
    }

    public <VM extends BaseViewModel> VM getViewModel(Class<? extends BaseViewModel> viewModel) {
        ViewModelProvider.Factory factory = new ViewModelProvider.NewInstanceFactory();
        VM vm = (VM) new ViewModelProvider(this, factory).get(viewModel);
        vm.setActivity(this);
        return vm;
    }
}

File: Trainning_Kotlin/app/src/main/java/com/example/myapplication/base/BaseActivity.kt
package com.example.myapplication.base

import android.os.Bundle
import android.os.PersistableBundle
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentTransaction
import butterknife.ButterKnife
import com.example.myapplication.R

abstract class BaseActivity : AppCompatActivity() {
    private lateinit var transaction : FragmentTransaction

    override fun onCreate(savedInstanceState: Bundle?, persistentState: PersistableBundle?) {
        super.onCreate(savedInstanceState, persistentState)
        setContentView(getLayoutId())
        ButterKnife.bind(this)
        transaction = supportFragmentManager.beginTransaction()
        initAct()
    }

    abstract fun initAct()

    fun getLayoutId(): Int {
        return R.layout.layout_container
    }

    protected fun pushView(fragment : Fragment) {
        transaction.add(R.id.container_frame, fragment)
        transaction.commitAllowingStateLoss()
    }

//    fun <VM : BaseViewModel> getViewModel(viewModel : Class<? : BaseViewModel>) : VM{
//        var factory : ViewModelProvider.Factory = ViewModelProvider.NewInstanceFactory()
//        var vm : VM = ViewModelProvider(this, factory).get(viewModel)
//        return vm
//    }

}